### ISSUE TITLE
No use case for this filter

### ISSUE THREAD

**Issue by vigie** [mgechev/angular-immutable#2]
Hi Minko

I am following up from my comments on your excellent blog series about using immutables with angular.

Given the current implementation of immutable.js and angular, I don't see how any performance benefit can be realized from the immutable filter in this repo. I understand there is performance benefit when using immutables as the expr in a $watch(expr, listener, false), due to the fast reference checking. 

But the point of this filter is to return an equivalent native datastrucutre from the immutable using the toJS() method. The object returned is always a new object. Because of this, if we were to use this filter in an expr to $watch(expr, listener, false), we would get many false positives and redundant invocations of listener (as you know the $digest is constantly running and evaluating expr even when expr isn't changing). 

If used with $watchCollection, such as in `<li ng-repeat="item in list | immutable">`, we get no performance benefit as watchCollection doesn't compare references, but always looks for changes in the collection regardless.

In fact, using this filter can only degrade performance, because of the garbage collection overhead you describe in your blog.

Please correct me if I'm understanding this wrong, and thanks for the research.
Matt

**Comment by haffmaestro** [comment]
Hi @mgechev , 
This was my experience as well, when using data structures more complicated than list of integers, say lists of immutable maps, this causes infinite digest loops.

However, as we were already really committed to using immutable data, mostly not for the performance increase but because of other benefits, we created a set of other angular helpers to better support this.

@vigie an implementation that supports more than lists with integers is available here [Immutable Angular](https://github.com/PaySavvy/immutable-angular). It uses a memoizing function to return the same object if the input is the same, the memoizing function uses an LRU cache so it shouldnt get too big.  

However, as far as I'm understanding it will still use $watchCollection, I dont know quite how to mitigate this, this is only true for the mutable filter, contained in the Immutable Angular package is also a directive that uses $watch(expr, listener, false) to be used for purposes when needing a mutable collection that is bound to the scope.  

@mgechev Thanks for the excellent blog posts series and the research, I learned a lot reading it!

**Comment by vigie** [comment]
@haffmaestro Thanks for the link, clever use of memoization in the filter. Agree it doesn't solve the watchCollection problem but at least the false positives problem is corrected. Directive also looks interesting.

**Comment by mgechev** [comment]
Outch! @vigie thanks for pointing this out!

The [initial implementation was a simple directive](https://github.com/mgechev/angular-immutable/blob/5e107cd45488ce1c3e6c172a261328d2bd96e3f5/lib/immutable.js), later [we decided](http://blog.mgechev.com/2015/03/02/immutability-in-angularjs-immutablejs/#comment-1889198535) to change it to a filter because it is more readable. However, most of the benchmarks I did weren't with watchers, which use the `immutable` filter and didn't profiled this further.

@haffmaestro you are using immutable data structures for building flux Angular applications?

I will restore the previous implementation and add link to @haffmaestro's repo in the readme.
